### PR TITLE
[SU-229][SU-231] New file browser part 2: directory tree

### PR DESCRIPTION
**Pull request by nawatts**
This is part 2 ([part 1](https://github.com/DataBiosphere/terra-ui/pull/3468)) of adding a new file browser to replace the files section in the workspace Data tab.

This adds a directory tree and a very basic files list (improved in [part 3](https://github.com/DataBiosphere/terra-ui/pull/3487)). Following PRs will add features like uploading and deleting files, etc.

To access the new file browser, enable the "Workspace Files Browser" feature preview (from http://localhost:3000/#feature-preview) and use the files button in the right sidebar.

Screen Shot 2022-11-03 at 11 01 37 AM

To try pagination, add a small `pageSize` parameter to the GCSFileBrowserBackend in WorkspaceFiles.ts. For example,
```ts
GCSFileBrowserBackend({ bucket: bucketName, project: googleProject, pageSize: 5 })
```

### CODE REVIEW

File: src/components/file-browser/FileBrowserDirectories.ts
Comment (cahrens) [inline src/components/file-browser/FileBrowserDirectories.ts]: Note that Links will become Buttons if they don't have an href defined.

**Comment by cahrens** [comment]
@nawatts I'd like to take a look at this from an a11y perspective when it's ready and you have a deployment up.

File: src/components/file-browser/FileBrowserDirectories.ts
Comment (nawatts) [reply src/components/file-browser/FileBrowserDirectories.ts]: I know, but Link seems to be our component for a text only button.

Comment (cahrens) [reply src/components/file-browser/FileBrowserDirectories.ts]: Clickable doesn't work? I ask because this caused confusion on a recent PR of ours-- the author assumed that creating a Link would result in an anchor tag.

Comment (nawatts) [reply src/components/file-browser/FileBrowserDirectories.ts]: That would lose the visual styles on Link.
https://github.com/DataBiosphere/terra-ui/blob/c31123e6328f56789203ba3a8f31e60786e61446/src/components/common.js#L95-L100

Comment (cahrens) [reply src/components/file-browser/FileBrowserDirectories.ts]: Thanks, I now think I have to go back and fix something. :)

File: src/pages/workspaces/workspace/analysis/ContextBar.js
Comment (cahrens) [inline src/pages/workspaces/workspace/analysis/ContextBar.js]: This is enabled for Azure workspaces, but results in an error.

**Comment by cahrens** [comment]
Should this be within a "Files" tab like the other Workspace "Views"? It feels like it's disconnected.

image

vs.

image

**Comment by cahrens** [comment]
I'm having trouble interacting with the tree view via the keyboard-- expand and collapsing elements. I think it's because your focus can move into the list item element, at which point toggling doesn't do anything.

I would expect to be able to expand collapse tree items with the arrow keys (like how it works in Finder), and move up and down with arrow keys. I don't know if we have any place in Terra that behaves in this way though.

File: src/pages/workspaces/workspace/analysis/ContextBar.js
Comment (cahrens) [reply src/pages/workspaces/workspace/analysis/ContextBar.js]: Also, should this be behind a feature flag?

**Comment by cahrens** [comment]
This role will probably help with getting the more natural keyboard interactions: https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/treeitem_role

File: src/pages/workspaces/workspace/analysis/ContextBar.js
Comment (nawatts) [reply src/pages/workspaces/workspace/analysis/ContextBar.js]: Oops, didn't consider Azure workspaces. And yes, I forgot to add the feature flag.

**Comment by nawatts** [comment]
> Should this be within a "Files" tab like the other Workspace "Views"? It feels like it's disconnected.

The button in the sidebar vs a tab was UX's decision. I've been thinking of it as the sidebar views are for more direct access to cloud resources while the tabs are Terra abstractions.

File: src/components/file-browser/FileBrowserDirectories.ts
Comment (aherbst-broad) [inline src/components/file-browser/FileBrowserDirectories.ts]: I like that we're abstracting the file browser behind this interface 👍

File: src/components/file-browser/FileBrowserDirectories.ts
Comment (aherbst-broad) [inline src/components/file-browser/FileBrowserDirectories.ts]: I'm curious, do we get enums with typescript, and if so would that be something to leverage for these sorts of comparisons?

**Review by aherbst-broad — APPROVED** [review]
I played with this locally, looking good.

File: src/components/file-browser/FileBrowserDirectories.ts
Comment (nawatts) [reply src/components/file-browser/FileBrowserDirectories.ts]: TypeScript does have [enums](https://www.typescriptlang.org/docs/handbook/enums.html), but we've elected to prefer [unions of literals](https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#literal-types) over them. For example:
```ts
const Status = 'Loading' | 'Ready' | 'Error'

if (status === 'Loading') { ... }
```
vs
```ts
enum Status {
  Loading,
  Ready,
  Error,
}

if (status === Status.Loading) { ... }
```

See https://www.bam.tech/article/should-you-use-enums-or-union-types-in-typescript for more information on the pros and cons of each.

We do get type checking here. For example, if I change this line to `if (status === 'error')` then TypeScript complains:
> This condition will always return 'false' since the types '"Loading" | "Ready" | "Error"' and '"error"' have no overlap.

**Review by MatthewBemis — APPROVED** [review]
Tested locally, code looks good. :+1: pending putting this behind a feature flag

File: src/pages/workspaces/workspace/analysis/ContextBar.js
Comment (cahrens) [reply src/pages/workspaces/workspace/analysis/ContextBar.js]: Remember to disable on Azure as well as the feature flag… and let me know if you figure out how to make the keyboard navigation more natural. These roles likely would help: https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/treeitem_role.

Comment (nawatts) [reply src/pages/workspaces/workspace/analysis/ContextBar.js]: Thanks for the reminder. I'm working on updating this to use tree roles and add keyboard nav.

**Comment by nawatts** [comment]
Updated the directory tree to use the tree role. Tested with VoiceOver in Chrome and Safari. It's not perfect, but much better than it was. Still need to add the feature flag and remove the link for Azure workspaces.

**Comment by cahrens** [comment]
Thanks @nawatts, this was a lot of work, and the way the left and right keys work to expand/collapse items definitely feels more natural now! I did hit this issue though after I collapsed the tree:

image

Screenreader support still is problematic, especially since contents load dynamically when a node is expanded. There is no feedback about what is going on, nor is the collapsed state getting reliably read out to me. I wouldn't say that is necessarily a blocker though, as realistically this whole tree view is going to be really challenging for a screen-reader user (how to they know what contents are changing on the right-hand side when focus changes in the tree?). It actually makes me wonder if there is some 3rd party library that we should be using for an accessible tree view.

**Comment by cahrens** [comment]
The duplication of "Files" feels odd here.

image

**Comment by cahrens** [comment]
@nawatts did you happen to look at https://dgreene1.github.io/react-accessible-treeview/? It's the first thing that came up when I searched… don't know what else is out there.

It does support asynchronous loading, which announces loading and loading complete when used with a screen reader. https://dgreene1.github.io/react-accessible-treeview/docs/examples-MultiSelectCheckboxAsync

**Comment by nawatts** [comment]
> Screenreader support still is problematic, especially since contents load dynamically when a node is expanded. There is no feedback about what is going on, nor is the collapsed state getting reliably read out to me. I wouldn't say that is necessarily a blocker though, as realistically this whole tree view is going to be really challenging for a screen-reader user (how to they know what contents are changing on the right-hand side when focus changes in the tree?).

Added screen reader announcements for loading/loaded based on https://dgreene1.github.io/react-accessible-treeview/docs/examples-MultiSelectCheckboxAsync. Also changed the visual loading indicator to a spinner instead of a "Loading" message.

As for the right panel, I was thinking of moving focus there when a directory was selected.

> The duplication of "Files" feels odd here.

Needed some label for the root path (''). Replaced "Files" with "Workspace bucket".

> I did hit this issue though after I collapsed the tree:

Where are you seeing that axe warning? I'm only seeing complaints about color contrast.

**Review by MatthewBemis — APPROVED** [review]

**Comment by cahrens** [comment]
To see the axe error (I'm using Axe development tools on your PR site), expand some of the nodes that load content, then collapse them. It is always reproducible for me.

image

**Comment by nawatts** [comment]
Ah, found it. It occurs when the tree item with keyboard focus is in the collapsed subtree. Then the aria-activedescendant attribute on the tree refers to an element that doesn't exist.

**Comment by cahrens** [comment]
I really like the aria live regions for loading state! And I like the spinner much more than the "Loading" subitem-- thanks.

The only wish I still have is that for some reason the screen reader is not consistently announcing things as you try to navigate back out of child directories (including collapsing items), so it can be somewhat confusing to figure out where you are. But possibly screen reader folks know special ways of interacting with trees? Not sure if there's something actionable here for you.

I'll start taking a look at the code this afternoon, between meetings.

**Comment by nawatts** [comment]
> or some reason the screen reader is not consistently announcing things as you try to navigate back out of child directories (including collapsing items), so it can be somewhat confusing to figure out where you are

Yeah, I found that to be the case even on some of the library demos, such as https://dgreene1.github.io/react-accessible-treeview/docs/examples-DirectoryTree.

**Comment by cahrens** [comment]
It would be good to have a tooltip on the folder icon in the context bar to indicate what it is.

It might be nice to put focus on the top tree node when you select the icon to open the Files view. Or perhaps hold off on that… I feel like UX needs to sort out how navigation between these views is supposed to work. The mix of breadcrumbs, tabs, and the context bar is pretty confusing… it doesn't seem clear how to get back to the tabbed view once you have gone into files. Who is the UX person you are working with?

**Comment by nawatts** [comment]
> I feel like UX needs to sort out how navigation between these views is supposed to work. The mix of breadcrumbs, tabs, and the context bar is pretty confusing… it doesn't seem clear how to get back to the tabbed view once you have gone into files. Who is the UX person you are working with?

@joyliu10

**Comment by kyuksel** [comment]
> > Screenreader support still is problematic, especially since contents load dynamically when a node is expanded. There is no feedback about what is going on, nor is the collapsed state getting reliably read out to me. I wouldn't say that is necessarily a blocker though, as realistically this whole tree view is going to be really challenging for a screen-reader user (how to they know what contents are changing on the right-hand side when focus changes in the tree?).
> 
> Added screen reader announcements for loading/loaded based on https://dgreene1.github.io/react-accessible-treeview/docs/examples-MultiSelectCheckboxAsync. Also changed the visual loading indicator to a spinner instead of a "Loading" message.

Does the PR address [SUP-777](https://broadworkbench.atlassian.net/browse/SUP-777) too then?

**Comment by cahrens** [comment]
Does the PR address [SUP-777](https://broadworkbench.atlassian.net/browse/SUP-777) too then?

No, @kyuksel. This PR is just about the tree view for workspace data (loading when a node is expanded). SUP-777 is about our general loading mechanism.

File: src/pages/workspaces/workspace/WorkspaceFiles.ts
Comment (cahrens) [inline src/pages/workspaces/workspace/WorkspaceFiles.ts]: I know we use `forwardRefWithName` all over the place, but I can't figure out what it actually does (and there is no documentation). Do you happen to know?

File: src/pages/workspaces/workspace/WorkspaceFiles.ts
Comment (cahrens) [inline src/pages/workspaces/workspace/WorkspaceFiles.ts]: Can you name this file just "Files.ts"? The workspace part is already well implied by the directories, and it will help with CODEOWNERS as the workspaces team won't get flagged on changes to this file.

File: src/pages/workspaces/workspace/WorkspaceFiles.ts
Comment (nawatts) [reply src/pages/workspaces/workspace/WorkspaceFiles.ts]: Since we use anonymous arrow functions, React can't figure out a component name for something like:
```js
const Component = forwardRef((props, ref) => ...)
```

`forwardRefWithName` adds a name to the component function so that it is identifiable in the Components panel of the React dev tools.

These would have the same effect:
```js
const Component = forwardRefWithName('MyComponent', (props, ref) => ...)
```

```js
const Component = forwardRef(function MyComponent(props, ref) { ... })
```

Comment (nawatts) [reply src/pages/workspaces/workspace/WorkspaceFiles.ts]: As for why it's needed here (despite `ref` being unused), WorkspaceContainer attaches a ref to its wrapped component. Without `forwardRef` here, React throws a warning: "Function components cannot be given refs. Attempts to access this ref will fail. Did you mean to use React.forwardRef()?"

https://github.com/DataBiosphere/terra-ui/blob/448e871bde8e429ed1092b27969b5ee97d8df34c/src/pages/workspaces/workspace/WorkspaceContainer.js#L417

File: src/components/file-browser/FileBrowser.ts
Comment (cahrens) [inline src/components/file-browser/FileBrowser.ts]: I don't understand-- why is there both a `FileBrowserDirectories` and a `FileBrowserFiles` (down at line 49)? I am wondering if some documentation would help make it clearer how everything fits together.

Comment (nawatts) [reply src/components/file-browser/FileBrowser.ts]: FileBrowserDirectories is the left panel with the directory tree. FileBrowserFiles is the right panel with the files in the selected directory.

Comment (cahrens) [reply src/components/file-browser/FileBrowser.ts]: Ah, what was confusing me is that a directory includes files…. I thought FileBrowserFiles was like a child node inside of a directory.

Comment (nawatts) [reply src/components/file-browser/FileBrowser.ts]: I could rename to something like FileBrowserDirectoryTree and FileBrowserFilesInDirectory?

Comment (cahrens) [reply src/components/file-browser/FileBrowser.ts]: Maybe just `DirectoryTree` and `FilesInDirectory` or `DirectoryContents`? Everything is already in `file-browser`, so I wonder if if it's really useful to prefix every file with FileBrowser.

File: src/components/file-browser/FileBrowserDirectories.test.ts
Comment (cahrens) [inline src/components/file-browser/FileBrowserDirectories.test.ts]: It might also be good to add in an axe check to verify aria attributes/ids. 

```
expect(await axe(container)).toHaveNoViolations()
```

File: src/components/file-browser/FileBrowserDirectories.ts
Comment (cahrens) [inline src/components/file-browser/FileBrowserDirectories.ts]: Why are there no types on these?

File: src/components/file-browser/FileBrowserDirectories.ts
Comment (cahrens) [inline src/components/file-browser/FileBrowserDirectories.ts]: Should status be an enum?

File: src/components/file-browser/FileBrowserDirectories.ts
Comment (cahrens) [inline src/components/file-browser/FileBrowserDirectories.ts]: It seems like we should define `#000` as a variable (at least within this file).

File: src/components/file-browser/FileBrowser.ts
Comment (nawatts) [reply src/components/file-browser/FileBrowser.ts]: 👍  I tend to name files for navigation via fuzzy finder. In other parts of the code, we have four files all named List.js.

File: src/components/file-browser/FileBrowserDirectories.ts
Comment (nawatts) [reply src/components/file-browser/FileBrowserDirectories.ts]: They are known from the return type of `useDirectoriesInDirectory`.

File: src/components/file-browser/FileBrowserDirectories.ts
Comment (nawatts) [reply src/components/file-browser/FileBrowserDirectories.ts]: We decided to use unions of literals instead of enums. `status` is type checked as one of "Loading", "Ready" or "Error".

Comment (cahrens) [reply src/components/file-browser/FileBrowserDirectories.ts]: Oh, can you point me to where the status literals are?

File: src/components/file-browser/FileBrowser.ts
Comment (cahrens) [reply src/components/file-browser/FileBrowser.ts]: I can see that, but here "FileBrowser" makes me think very specifically of the directory tree, which added to my confusion.

File: src/components/file-browser/FileBrowserDirectories.ts
Comment (nawatts) [reply src/components/file-browser/FileBrowserDirectories.ts]: [src/libs/type-utils/LoadedState.ts](https://github.com/DataBiosphere/terra-ui/blob/su-229-file-browser-2/src/libs/type-utils/LoadedState.ts).

`useDirectoriesInDirectory` uses `useIncrementalResponse`. The type of the `state` property returned from `useIncrementalResponse` is based on `LoadedState`.

https://github.com/DataBiosphere/terra-ui/blob/1f75600e7cb8165e47a5391b876c1dedd47f2d87/src/components/file-browser/file-browser-hooks.ts#L16-L24

https://github.com/DataBiosphere/terra-ui/blob/1f75600e7cb8165e47a5391b876c1dedd47f2d87/src/libs/ajax/incremental-response/useIncrementalResponse.ts#L9-L17

File: src/components/file-browser/FileBrowserDirectories.ts
Comment (cahrens) [reply src/components/file-browser/FileBrowserDirectories.ts]: This is something I'm trying to get used to… that the types aren't repeated so when you are reading the code outside of an IDE it's not easy to tell what the types are.

File: src/components/file-browser/FileBrowserDirectories.ts
Comment (cahrens) [inline src/components/file-browser/FileBrowserDirectories.ts]: It would be nice if the keyboard focus went to the first loaded node when "Load Next Page" completes. As it is, you have to navigate back up and understand what was there before you triggered the action.

File: src/components/file-browser/FileBrowserDirectories.ts
Comment (cahrens) [reply src/components/file-browser/FileBrowserDirectories.ts]: Interesting… this will take me some time to get used to also!

File: src/components/file-browser/FileBrowserDirectories.ts
Comment (nawatts) [reply src/components/file-browser/FileBrowserDirectories.ts]: Added in https://github.com/DataBiosphere/terra-ui/pull/3469/commits/2b09e10ee894418346ff07931134e02ba78f8a45.

File: src/components/file-browser/FileBrowserDirectories.test.ts
Comment (nawatts) [reply src/components/file-browser/FileBrowserDirectories.test.ts]: Added in https://github.com/DataBiosphere/terra-ui/pull/3469/commits/ca23519197059376718c5a5107f814af4b2ff00c.

File: src/components/file-browser/FileBrowser.ts
Comment (nawatts) [reply src/components/file-browser/FileBrowser.ts]: Renamed in https://github.com/DataBiosphere/terra-ui/pull/3469/commits/7532cdb1ae394e50958754ce90ee32b56a542f76.

File: src/components/file-browser/FileBrowserDirectories.ts
Comment (nawatts) [reply src/components/file-browser/FileBrowserDirectories.ts]: Done in https://github.com/DataBiosphere/terra-ui/pull/3469/commits/9ddd07e5eacaf47087d534a45b20cdc276230049.

**Review by cahrens — APPROVED** [review]

**Pull request merged**
